File: voltdb/procs/NextInSequence.java
import org.voltdb.*;

@ProcInfo (
  partitionInfo = "sequences.name: 0",
  singlePartition = true
)

public class NextInSequence extends VoltProcedure {
  public final SQLStmt selectSQL =
    new SQLStmt( "SELECT nextval FROM sequences WHERE name=?");

  public final SQLStmt initSQL =
    new SQLStmt( "INSERT INTO sequences (name, nextval) VALUES (?, 2)");

  public final SQLStmt incrementSQL =
    new SQLStmt( "UPDATE sequences SET nextval=nextval+1 WHERE name=?");

  public long run(String name) {
    voltQueueSQL(selectSQL, name);
    VoltTable res1 = voltExecuteSQL()[0];

    if(res1.getRowCount() < 1) {
      voltQueueSQL(initSQL, name);
      voltExecuteSQL();
      return 1;
    }
    else {
      voltQueueSQL(incrementSQL, name);
      voltExecuteSQL();
      return res1.fetchRow(0).getLong(0);
    }
  }
}

File: voltdb/procs/next_in_sequence.clj
(ns next-in-sequence)
(require 'u)

(defn statements []
  {"next-in-sequence-select"
     (u/stmt "SELECT nextval FROM sequences WHERE name=?")
   "next-in-sequence-init"
     (u/stmt "INSERT INTO sequences (name, nextval) VALUES (?, 2)")
   "next-in-sequence-incr"
     (u/stmt "UPDATE sequences SET nextval=nextval+1 WHERE name=?")})

(defn run [this name]
  ; upsert next long in sequence
  (u/queue this "next-in-sequence-select" name)
  (let [res1 (nth (u/execute this) 0)]
    (if (< (.getRowCount res1) 1)
      (do (u/queue this "next-in-sequence-init" name)
          (u/execute this)
          1)
      (do (u/queue this "next-in-sequence-incr" name)
          (u/execute this)
          (.getLong (.fetchRow res1 0) 0)))))
